File: Hash Tables Maps/two_sum.py

"""
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.
#? Time O(N): We do a loop over array of size n
#? Space O(N): We create a map to hold a value alongside the index
"""


# def two_sum(nums, target):
#     my_dict = {}
#     for num in nums:
#         goal = target- num
#         if goal in my_dict:
#             return (my_dict[goal], nums.index(num))
#         my_dict[num]= nums.index(num)

# #* Another solution:
# def two_sum(nums,target):
#         m = {}
#         n = len(nums)
#         for i in range(0,n):
#             goal = target - nums[i]
#             if(goal in m):
#                 return [m[goal], i]
#             m[nums[i]] = i

# #*Another way
# def twoSum(nums, target):
# 	num_to_index = {}
# 	for i, num in enumerate(nums):
# 		if target - num in num_to_index:
# 			return (i, num_to_index[target - num])
# 		num_to_index[num] = i


# *third attempt

# def two_sum(nums, target):
#     nums_to_index = {}
#     for i in range(0,len(nums)):
#         goal = target - nums[i]
#         if goal in nums_to_index:
#             return nums_to_index[goal], i
#         nums_to_index[nums[i]] = i

# *Whiteboard Solution

# def two_sum(nums, target):
#     nums_to_index = {}

#     for i in range(0, len(nums)):
#         goal = target - nums[i]
#         if goal in nums_to_index:
#             return nums_to_index[goal], i
#         nums_to_index[nums[i]]= i

# nums= [3,3]
# target = 6

# result = two_sum(nums, target)

#print (f"The numbers that add up to {target} can be found in indices {result[0]} and {result[1]}.")


# * 2022 Solution


# def two_sum(nums, target):

#     nums_to_index = {}

#     for i in range (0, len(nums)):
#         goal = target - nums[i]
#         if goal in nums_to_index:
#             return nums_to_index[goal], i
#         nums_to_index[nums[i]] = i

def two_sum(nums, target):
    nums_to_index = {}

    for i in range(0, len(nums)):
        goal = target - nums[i]
        if goal in nums_to_index:
            return nums_to_index[goal], i
        nums_to_index[nums[i]] = i


nums = [5, 6, 4, 7, 2, 9]
target = 6

result = two_sum(nums, target)

print(
    f"The numbers that add up to {target} can be found in indices {result[0]} and {result[1]}.")

File: Hash Tables Maps/twoSum.js
/**
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

LeetCode: https://leetcode.com/problems/two-sum/
LeetCode Number: 1
Difficulty: Easy
Topic: Hash Tables, Maps
Company: Amazon, Apple, Facebook, Google, Microsoft, Uber, Yahoo

#? Time O(N): We do a loop over array of size n
#? Space O(N): We create a map to hold a value alongside the index
*/

// function twoSum(nums, target) {
//     let map = new Map();
//     for (let i = 0; i < nums.length; i++) {
//         let complement = target - nums[i];
//         if (map.has(complement)) {
//             return [map.get(complement), i];
//         }
//         map.set(nums[i], i);
//     }
//     return null;
// }

var twoSum = function(nums, target) {
    let numToIndex = {};
    for (let i = 0; i < nums.length; i++) {
        let goal = target - nums[i];
        if (goal in numToIndex) {
            return [i, numToIndex[goal]];
        }
        numToIndex[nums[i]] = i;
    }
    return null;
};

var nums = [3, 3];
var target = 6;

var result = twoSum(nums, target);

console.log(twoSum([2, 7, 11, 15], 9)); // [0, 1]
console.log(twoSum([3, 2, 4], 6)); // [1, 2]
console.log(twoSum([3, 3], 6)); // [0, 1]
console.log(twoSum([3, 2, 3], 20)); // null

var twoSum2 = function(nums, target) {
    numToIndex = {};
    for (let i = 0; i < nums.length; i++) {
        let goal = target - nums[i];
        if (goal in numToIndex) {
            return ([i, numToIndex[goal]]);
        }
        numToIndex[nums[i]] = i;
    }
    return null;
};

var nums = [3, 4, 5];
var target = 8;

var result = twoSum2(nums, target);

console.log(result);